### DISCUSSION TITLE
STM32F411 Bluetooth

### DISCUSSION THREAD

**Discussion by Profd79** [grblHAL/Plugins_Bluetooth#2] [Q&A]
For this board STM32F411 which pins should be used for Bluetooth ?

**Comment by terjeio** [comment]
The plugin will use the first free serial port and STATE pin. I recently have added $-setting 385 for the STATE pin to use, set it to the port number for the pin as listed in by the $pins command. I have to add documentation for that... The state pin selected must support the change interrupt type (firing on both going low and high), currently you have to try the ports available if the plugin complains about the selected pin.

**Comment by Profd79** [comment]
pin map for my board
$PINS < [PIN:PB6,Reset]
[PIN:PB7,Feed hold]
[PIN:PB8,Cycle start]
[PIN:PB15,Probe]
[PIN:PB12,X limit min]
[PIN:PB13,Y limit min]
[PIN:PB14,Z limit min]
[PIN:PA0,X step]
[PIN:PA2,Y step]
[PIN:PA4,Z step]
[PIN:PA1,X dir]
[PIN:PA3,Y dir]
[PIN:PA5,Z dir]
[PIN:PB0,Steppers enable]
[PIN:PB1,Spindle on]
[PIN:PB2,Spindle direction]
[PIN:PC15,Flood]
[PIN:PC14,Mist]
[PIN:PB10,SCK]
[PIN:PB11,SDA]
[PIN:PA10,RX,Bluetooth]
[PIN:PA9,TX,Bluetooth]
[PIN:PA8,Spindle PWM]

but on this pins 
[PIN:PA10,RX,Bluetooth]
[PIN:PA9,TX,Bluetooth]

no signal.
I try connect HC05 and usb to uart converter to listenn uart commands nothing .....
$377 < error:53 this command return error.

**Reply by terjeio** [reply]
There is no aux input for the STATE signal, without that the plugin cannot be used.
Which board map do you use? It might be possible to assign a motor driver pin depending on the number of axes you have configured.

> I try connect HC05 and usb to uart converter to listenn uart commands nothing .....

The module is designed to take full control when a client connects and for that a compatible aux input is required.
I should add a mode where a client is just listening to replicated output from the master connection?

> $377 < error:53 this command return error.

$377 is only enabled when the required aux input is available.

**Comment by Profd79** [comment]
i try this two config blackpill_map.h or mini_blackpill_map.h and dont use A Axis motor.

**Reply by terjeio** [reply]
Ok, I guess you are lucky then. Add these lines to the end of either of these maps:
```
#define AUXINPUT0_PORT          GPIOB
#define AUXINPUT0_PIN           15
#define AUXOUTPUT0_PORT         GPIOA
#define AUXOUTPUT0_PIN          7
```
and change probe mapping to
```
#define PROBE_PORT              GPIOA
#define PROBE_PIN               6
```
and you will get one aux input and one output.

Note that pin 6 and 7 cannot be used as input for the STATE signal as they are not interrupt capable.

**Reply by terjeio** [reply]
Another option is to use this configuration, and is what I will add as default to those maps:
```
#if N_ABC_MOTORS == 0
#if !SAFETY_DOOR_ENABLE
#define AUXINPUT0_PORT          GPIOB
#define AUXINPUT0_PIN           9
#define AUXOUTPUT1_PORT         GPIOA
#define AUXOUTPUT1_PIN          6
#else
#define AUXINPUT0_PORT          GPIOA
#define AUXINPUT0_PIN           6
#endif
#define AUXOUTPUT0_PORT         GPIOA
#define AUXOUTPUT0_PIN          7
#endif
```

**Comment by Profd79** [comment]
i add tihis line to config 
but the compile warning
Src\driver.c: In function 'EXTI15_10_IRQHandler':
Src\driver.c:2444:13: warning: implicit declaration of function 'ioports_event'; did you mean 'ioport_claim'? [-Wimplicit-function-declaration]
             ioports_event(ifg & aux_irq);
             ^~~~~~~~~~~~~
             ioport_claim

i resold this pins to HC05 and try

**Reply by terjeio** [reply]
Oops, add a line with
#define HAS_IOPORTS
to the board map somewhere.

**Comment by Profd79** [comment]
ok no warning compiling

i got this pins

$PINS < [PIN:PB6,Reset]
[PIN:PB7,Feed hold]
[PIN:PB8,Cycle start]
[PIN:PA6,Probe]
[PIN:PB12,X limit min]
[PIN:PB13,Y limit min]
[PIN:PB14,Z limit min]
[PIN:PB15,Aux input 0,HC-05 STATE]
[PIN:PA0,X step]
[PIN:PA2,Y step]
[PIN:PA4,Z step]
[PIN:PA1,X dir]
[PIN:PA3,Y dir]
[PIN:PA5,Z dir]
[PIN:PB0,Steppers enable]
[PIN:PB1,Spindle on]
[PIN:PB2,Spindle direction]
[PIN:PC15,Flood]
[PIN:PC14,Mist]
[PIN:PA7,Aux out 0,P0]
[PIN:PB10,SCK]
[PIN:PB11,SDA]
[PIN:PA10,RX,Bluetooth]
[PIN:PA9,TX,Bluetooth]
[PIN:PA8,Spindle PWM]
ok

i must  connect HC05 state pin to this [PIN:PB15,Aux input 0,HC-05 STATE] ?
and TX RX to  [PIN:PA10,RX,Bluetooth] [PIN:PA9,TX,Bluetooth]


an now $377=1 < ok             comand $377 work

**Reply by terjeio** [reply]
Yes you have to connect the pins as shown, RX/TX is crossed: RX > TX, TX > RX. If the module is not configured for 115000 baud operation you have to either somehow do that manually or use the [auto configure](https://github.com/grblHAL/Plugins_Bluetooth#auto-configure) method.

**Comment by Profd79** [comment]
I change PB15 nj PB5
#define AUXINPUT0_PORT          GPIOB
#define AUXINPUT0_PIN           5

cinfirm Bluetooth work.

thank you very much.

P.S. It is posible to write Servo plugin ?

**Reply by terjeio** [reply]
> P.S. It is posible to write Servo plugin ?

For monitoring motor status signals yes I believe so. Or are you thinking about something else?

Tip: since you have modified the board map file you may rename it to _my_machine_map.h_ and enable `BOARD_MY_MACHINE` in _my_machine.h_. This way your changes will not be overwritten by an update.

**Comment by ghost** [comment]
Hi terjeio,

I am custom building a board with a 64 Pin STM32F401 MCU and I want to be able to auto switch from USB (PA2 & PA3 [USART_2]) to WiFi (ESP-01) connected to PB6 & PB7 (USART_1). Is it possible to use the VUSB (5v dropped to 3.3v) so that when there is 0v (pulled down) it will then provide a 3.3v output on one of the AUX Output pins to enable the ESP-01?

Cheers
George

**Reply by terjeio** [reply]
> Is it possible to use the VUSB (5v dropped to 3.3v) so that when there is 0v (pulled down) it will then provide a 3.3v output on one of the AUX Output pins to enable the ESP-01?

Yes. But would it not be simpler to use a transistor (MOSFET?) to switch power? Saves an output pin.
There may also be a register bit/associated interrupt for VUSB detection - check the datasheet. If so you may save an input pin as well.

**Reply by ghost** [reply]
Thanks terjeio.

So if I was using the ESP-01 WiFi module (which does not have a 'State' output) I would still lose an Aux Input (highest numbered) but potentially would not need to use it.

Also with the current grblHal firmware, if Bluetooth is enabled does it detect that there is no TX/RX signal on USART_2 (PA2 & PA3) to then change over to USART_1 (PB6 & PB7)?

I agree that a simple switch MOSFET will be a better option for the ESP-01 'Enable' input when the USB plug is disconnected.

Cheers
George

**Reply by terjeio** [reply]
> So if I was using the ESP-01 WiFi module (which does not have a 'State' output) I would still lose an Aux Input (highest numbered) but potentially would not need to use it.

Yes, unless you detect that the VUSB is present in some other way (if you are going to use that for the change over).

> Also with the current grblHal firmware, if Bluetooth is enabled does it detect that there is no TX/RX signal on USART_2 (PA2 & PA3) to then change over to USART_1 (PB6 & PB7)?

No, the state signal is used for the change over. If using RX data for detecting that a client is active a timeout has to be added. I am not sure that is a good idea - at least my sender stop sending real time report requests when some of the UI tabs are activated...

BTW it might be possible to customize the ESP-01 code to provide a state output?

**Reply by ghost** [reply]
Cool.

I can use the VUSB (via a MOSFET) to go high (3.3V) when the USB plug is removed which could also go to the AUX Input as the changeover (assuming that the AUX Input is active high).

Cheers
George
